### PR TITLE
Clarifies that processProposal may be called for set of transactions different from the one returned in the preceding prepareProposal

### PR DESCRIPTION
**Pull request by lasarojc**
If a proposer fails after calling prepareProposal and before calling processProposal, then the following may happen upon restarting:
- if failed before signing another message, then will invoke prepareProposal again, sign a new block, probably empty, and propose it;
- if failed after signing a proposal but before writing the  proposal message into the WAL, then will invoke prepareProposal and produce a new, probably empty block, [fail to sign it](https://github.com/cometbft/cometbft/blob/2789a59a9cc61c6ea56a6b266eeadf0f26ca2456/consensus/state.go#L1221), and not invoke processProposal; prevote timeouts will ensure the CometBFT is not stuck;
- if failed after writing the proposal message to the WAL, then will invoke prepareProposal, produce a new, probably empty block, fail to sign it, and invoke processProposal with the block signed before crashing.

---

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [x] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [inline spec/abci/abci++_basic_concepts.md:215]: If the implementation of `PrepProp` and `ProcProp` are bounded in time, we don't introduce liveness problems (the algorithm is guaranteed to terminate), but performance problems (unwanted rounds).

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [inline spec/abci/abci++_basic_concepts.md:209]: Overall, I like this write-up more than the previous one.

Just on thing: you're introducing `ProcProp` in the reasoning, but the problem is way less serious with `ProcProp` than with `PrepProp`:

* At the beginning of a round, all non-proposers will schedule `TimeoutPropose` immediately
* however, with `ProcProp` the `TimeoutPropose` timeout is only scheduled after having received 2/3 of all prevotes. In practice, since the block tends to take a similar time to run at all validators, when a validator schedules `TimeoutPropose` it is likely to be about to finish its own `ProcProp` execution.

IOW, the _real_ problem is `PrepProp` and `TimeoutPropose`

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [inline spec/abci/abci++_basic_concepts.md:242]: Did you mean synchronously?
If you execute blocks asynchronously, then you can't reject the invalid blocks.
Or maybe I'm not fully understanding

Comment (sergio-mena) [reply spec/abci/abci++_basic_concepts.md:242]: Remember, what we achieve with immediate execution is the 100% guarantee that no TX included in a block will fail.
The app may choose to execute the block in parallel by replying `ACCEPT` right away, but then it cannot reject blocks based on its execution (in the current version of ABCI).

File: spec/abci/abci++_methods.md
Comment (sergio-mena) [inline spec/abci/abci++_methods.md:446]: Excellent!

File: spec/abci/abci++_comet_expected_behavior.md
Comment (sergio-mena) [inline spec/abci/abci++_comet_expected_behavior.md:160]: I think we need to adapt this as well

Comment (sergio-mena) [reply spec/abci/abci++_comet_expected_behavior.md:160]: Also, in the end, we agreed not to touch line 59 above, didn't we?
I think the final decision was to leave that line untouched (but not 100% sure, hence double-checking)

File: spec/abci/abci++_basic_concepts.md
Comment (lasarojc) [reply spec/abci/abci++_basic_concepts.md:209]: Good point, @sergio-mena. Let me rephrase it to remove ProcProp.

File: spec/abci/abci++_basic_concepts.md
Comment (lasarojc) [reply spec/abci/abci++_basic_concepts.md:242]: I really meant asynchronouusly, by replying ACCEPT right away. I modified the previous change to explicitly call out immediate execution (maybe a better name would be early execution) and the following change to discuss both immediate execution and optimistic/asynchronous/parallel execution.

File: spec/abci/abci++_comet_expected_behavior.md
Comment (lasarojc) [reply spec/abci/abci++_comet_expected_behavior.md:160]: I didn't change anything in this file because the grammar specifies the behavior in correct processes (that's why we agreed not to change line 59).
The main change hapened in the `example_scenarios` file, but I have now updated the text here. Let me know if I should revert.

File: spec/abci/abci++_basic_concepts.md
Comment (cason) [inline spec/abci/abci++_basic_concepts.md:222]: Can you break this line please?

Comment (cason) [reply spec/abci/abci++_basic_concepts.md:222]: Actually, the validators have to verify the block execution in `ProcessProposal`.

Comment (cason) [reply spec/abci/abci++_basic_concepts.md:222]: I didn't understand the second part of the sentence, starting from "to ensure that..."

File: spec/abci/abci++_basic_concepts.md
Comment (cason) [inline spec/abci/abci++_basic_concepts.md:254]:
```suggestion
that the proposed block was decided and `Commit` is invoked for it.
```

File: spec/abci/abci++_comet_expected_behavior.md
Comment (cason) [inline spec/abci/abci++_comet_expected_behavior.md:159]: Sorry, it is not this part that we have to change?

Comment (cason) [reply spec/abci/abci++_comet_expected_behavior.md:159]: If the node is the proposer of the current round, we call `PrepareProposal`.

If the proposer doesn't crash, `ProcessProposal` is called for the same (prepared) proposal.

If the proposer crashes and then recovers, there are some possibilities, we cannot ensure in the current implementation this match of prepare and process the same proposal.

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [reply spec/abci/abci++_basic_concepts.md]: > all non-proposers will schedule TimeoutPropose immediately

Actually, that's wrong as we saw on Fri (it's true in the arXiv paper, but not in the code)

**Review by cason — COMMENTED** [review]
I don't think this new redaction is really clarifying the situation mentioned in the title of this PR.

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [reply spec/abci/abci++_basic_concepts.md]: Ok I'll re-read in the next pass

File: spec/abci/abci++_comet_expected_behavior.md
Comment (lasarojc) [reply spec/abci/abci++_comet_expected_behavior.md:159]: It was changed. Let me try to make it more clear.

Comment (cason) [reply spec/abci/abci++_comet_expected_behavior.md:159]: Sorry, we had a concurrency problem here probably. Now I've read you preferred not update the wording here because the grammar says otherwise.

File: spec/abci/abci++_basic_concepts.md
Comment (lasarojc) [reply spec/abci/abci++_basic_concepts.md:222]: We need to agree on how to break lines, if at the 80 or 100 char mark or something else.
I prefer breaking into periods, even if it makes lines longer.

I didn't catch this comment.
> Actually, the validators have to verify the block execution in ProcessProposal.

Do you mean that I should say validate instead of execute?
I guess I just expanded "validate" to "execute a block (...) to ensure all transactions are executable in case the block is decided."

Comment (cason) [reply spec/abci/abci++_basic_concepts.md:222]: We can have two cases here:
1. Everyone executes the block in `ProcessProposal`, as a pre-execution that will actually be confirmed when the block is finally committed
2. The proposes execute the block on `PrepareProposal`, encodes somehow the results of the execution on the block itself, and non-proposers have to check whether the proposer's execution is valid, in `ProcessProposal`

I think we should discourage option 2, because it adds a lot of complexity and corner cases. My suggestion can be just to mention case 1 here.

Comment (lasarojc) [reply spec/abci/abci++_basic_concepts.md]: We need to agree on how to break lines, if at the 80 or 100 char mark or something else.
I prefer breaking into periods, even if it makes lines longer.

I didn't catch this comment.
> Actually, the validators have to verify the block execution in ProcessProposal.

Do you mean that I should say validate instead of execute?
I guess I just expanded "validate" to "execute a block (...) to ensure all transactions are executable in case the block is decided."

Comment (cason) [reply spec/abci/abci++_basic_concepts.md]: > to ensure that all transactions are executable in case the block is decided.

This part I really don't get: are we executing txs to ensure they can be executed? If we are executing them now, we don't execute them again upon Commit. Upon commit we only check whether the block that we previously executed is the same block that has been committed. If this is not the case, we execute the block on Commit replacing the (optimistic) execution we performed upon `ProcessProposal`.

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [reply spec/abci/abci++_basic_concepts.md:215]: > Ah, this is a very important information. What happens if the application doesn't reply on time? What is proposed?

CometBFT blocks until the application replies (similarly to `FinalizeBlock`). It's a **sync** ABCI call

File: spec/abci/abci++_basic_concepts.md
Comment (sergio-mena) [inline spec/abci/abci++_basic_concepts.md:255]:
```suggestion
The same is true to Applications that quickly accept blocks and execute the blocks optimistically in parallel with the remaining consensus steps to save time during `FinalizeBlock`; they must only apply state changes in `Commit`.
```

Thanks for improving the wording here

File: spec/abci/abci++_basic_concepts.md
Comment (cason) [reply spec/abci/abci++_basic_concepts.md:215]: So, timeout prevote does not apply here, as it is only triggered when the node receives 2/3+ votes that do not form a quorum.

File: spec/abci/abci++_comet_expected_behavior.md
Comment (sergio-mena) [inline spec/abci/abci++_comet_expected_behavior.md:159]:
```suggestion
These two always come together because once a proposal is prepared it is broadcast to all validators, including the proposer itself. In the absence of crashes, `ProcessProposal` is invoked with the proposal returned by `PrepareProposal`.
```

File: spec/abci/abci++_basic_concepts.md
Comment (cason) [reply spec/abci/abci++_basic_concepts.md:215]: > I still feel we shouldn't say that liveness is compromised (but performance is degraded). Thoughts?

I think it affects liveness somehow. The point is that timeouts are increasing, so eventually they become big enough so that to include the processing time involved on these calls.

Comment (lasarojc) [reply spec/abci/abci++_basic_concepts.md:215]: > > I still feel we shouldn't say that liveness is compromised (but performance is degraded). Thoughts?
> 
> I think it affects liveness somehow. The point is that timeouts are increasing, so eventually they become big enough so that to include the processing time involved on these calls.

Good point. At some point then proposals will be made before the other validators prevote nil and eventually we decide. I agree with you that it is not a liveness issue, but a performance one. A potentially very painful one, since the timeout values are reset for each new round.

File: spec/abci/abci++_methods.md
Comment (sergio-mena) [inline spec/abci/abci++_methods.md:446]:
```
      However, there is no guarantee that the values of `RequestProcessProposal.txs` and `RequestProcessProposal.hash` will be the same as returned in the preceding call to `PrepareProposal`.
```suggestion

Also, we may need to clarify that `RequestProcessProposal.hash` is not returned by `RequestPrepareProposal`, but calculated by CometBFT just after `PrepareProposal` returns.

**Review by sergio-mena — APPROVED** [review]
Thanks for this!

**Comment by sergio-mena** [comment]
Shall we backport this to `v0.38.x` and `v0.37.x`?

File: spec/abci/abci++_methods.md
Comment (lasarojc) [inline spec/abci/abci++_methods.md:445]: Considering the second bullet in the description, the following may be a better explanation.

```suggestion
    * `RequestProcessProposal` is also called at the proposer of a round. The reason for this is to
      inform the Application of the block header's hash, which is only calculated by CometBFT after the execution of `PrepareProposal`.
      Normally the call to `RequestProcessProposal` occurs right after the call to `RequestPrepareProposal` and
      `RequestProcessProposal` matches the block produced based on `ResponsePrepareProposal` (i.e.,
      `RequestPrepareProposal` equals `RequestProcessProposal.txs`).
      However, no such guarantee is made since, in the presence of failures, `RequestProcessProposal` may match
      `ResponsePrepareProposal` from an earlier invocation or `ProcessProposal` may not be invoked at all.
```

Comment (sergio-mena) [reply spec/abci/abci++_methods.md:445]: I would remove the sentence "the reason for this is to inform the Application of the block header's hash, which is only calculated by CometBFT after the execution of `PrepareProposal`."

Given the current situation, it only adds confusion IMO

**Comment by lasarojc** [comment]
> Shall we backport this to `v0.38.x` and `v0.37.x`?

I think we should. I will wait for @cason's opinion and tag accordingly before merging.

**Review by cason — APPROVED** [review]
I am not sure whether the new grammar is precise, in this case, as it allows for individual calls to the pair Prepare/ProcessProposal, which I am not sure that is the case, even in replay mode.

In any case, I believe that the changes made do, in particular, make clear that there is no a necessary bond between the proposal prepared and processed by a proposer, as it can process at most a proposal per round of consensus (despite crashes), but it may prepare more than one proposal per round.

I have left some comments that should not block merging this PR.

File: spec/abci/abci++_comet_expected_behavior.md
Comment (lasarojc) [reply spec/abci/abci++_comet_expected_behavior.md:262]: I agree it is confusing sometimes. If using GitHub's web interface, then it highlights where the differences are, but it is not perfect. In this case, there is really no change.

File: spec/abci/abci++_methods.md
Comment (lasarojc) [inline spec/abci/abci++_methods.md:443]:
```suggestion
```

File: spec/abci/abci++_comet_expected_behavior.md
Comment (cason) [reply spec/abci/abci++_comet_expected_behavior.md:262]: I was looking for a plugin for mardown diff, but I didn't find anything that can be loaded into GitHub...

File: spec/abci/abci++_comet_expected_behavior.md
Comment (lasarojc) [inline spec/abci/abci++_comet_expected_behavior.md:59]:
```suggestion
proposer            = *got-vote [prepare-proposal [process-proposal]] [extend]
```

File: spec/abci/abci++_comet_expected_behavior.md
Comment (lasarojc) [inline spec/abci/abci++_comet_expected_behavior.md:178]:
```suggestion
>proposer            = *got-vote [prepare-proposal [process-proposal]] [extend]
```

**Comment by lasarojc** [comment]
Thanks for the reviews. I've fixed the grammar to have `ProcessProposal` follow `PrepareProposal`.
Merging the changes.

**Pull request merged**
